### PR TITLE
[Google Drive] Support domain-wide delegation for syncing data

### PR DESCRIPTION
**Pull request by jedrazb**
## Closes https://github.com/elastic/connectors-python/issues/1508

This allows to sync all organization data (shared drives + all personal drives) without needing to share anything with a service account. 

Check "tooltip" for detailed description of new RCFs.

The service account needs to have added addditional scopes in google admin to make this work:
- `https://www.googleapis.com/auth/drive.readonly`
- `https://www.googleapis.com/auth/drive.metadata.readonly`

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [ ] (followup ticket) Considered corresponding documentation changes 
- [ ] (followup ticket) Contributed any configuration settings changes to the configuration reference

## Manual tests

| Domain-Wide delegation sync enabled | DLS enabled | Does it work?           |
|-------------------------------------|-------------|-------------------------|
| `False`                             | `False`     | ✅ (content sync)        |
| `False`                             | `True`      | ✅ (content and AC sync) |
| `True`                              | `False`     | ✅ (content sync)        |
| `True`                              | `True`      | ✅ (content and AC sync) |

## RCF Changes

You need to enter admin email only once:

**Case 1**: Domain-wide delegation sync enabled, admin email renders on top (regardless if DLS is enabled)

Screenshot 2023-09-22 at 11 40 38

**Case 2**: Domain-wide delegation sync disabled, admin email required for DLS 

Screenshot 2023-09-22 at 11 40 24

## Release Note

Added support for syncing data from Google Drive using domain-wide delegation of authority. This allows a service account to access user data on behalf of domain users.

### CODE REVIEW

**Comment by jedrazb** [comment]
Documentation changes on docs.elastic will be made after this gets merged

File: connectors/sources/google_drive.py [hunk 1]
```diff
@@ -515,8 +583,7 @@ def get_default_configuration(cls):
             },
         }
 
-    @cached_property
-    def google_drive_client(self):
+    def google_drive_client(self, impersonate_email=None):
```
Comment (timgrein) [inline connectors/sources/google_drive.py:586]: Maybe we can add a small comment explaining why this is not cached and why we need to reinstantiate?

File: connectors/sources/google_drive.py [hunk 2]
```diff
@@ -589,29 +672,85 @@ def _validate_google_workspace_admin_email(self):
...
+        if self._domain_wide_delegation_sync_enabled():
+            return self.configuration["google_workspace_admin_email_for_data_sync"]
+        elif self._dls_enabled():
+            return self.configuration["google_workspace_admin_email"]
+        else:
+            return None
+
+    def _google_google_workspace_email_for_shared_drives_sync(self):
+        """Get Google Workspace email for syncing shared drives"""
+        return self.configuration.get("google_workspace_email_for_shared_drives_sync")
```
Comment (timgrein) [inline connectors/sources/google_drive.py:752]: I don't know, if the comment adds any value here and it should probably be `_get_google[...]` and not `_google_google`?

```suggestion
    def _get_google_workspace_email_for_shared_drives_sync(self):
        return self.configuration.get("google_workspace_email_for_shared_drives_sync")
```

File: tests/sources/test_google_drive.py [hunk 3]
```diff
@@ -26,13 +27,14 @@
...
-async def create_gdrive_source():
+async def create_gdrive_source(**kwargs):
     async with create_source(
         GoogleDriveDataSource,
         service_account_credentials=SERVICE_ACCOUNT_CREDENTIALS,
         use_document_level_security=False,
-        google_workspace_admin_email="admin@your-organization.com",
+        **kwargs
     ) as source:
+        # source.google_drive_client = mock.MagicMock()
```
Comment (timgrein) [inline tests/sources/test_google_drive.py:37]: Should we remove that line?

**Review by timgrein — DISMISSED** [review]
Only some nits, nothing blocking.

Great work and really detailed PR description 👏

**Review by vidok — APPROVED** [review]
LGTM

**Pull request merged**